v1.3.X miner hanging issues

Please comment in this issue if you experience any issues with the v1.3.X version of the miner hanging after it has been running for a while.

Please include:
What OS (Windows, Linux, MacOs)
What Arch (386, amd64, ARM, ARM64)
The log output for your miner. If you are uncomfortable including in this issue, please DM me the logs on Discord.

CentOS 7 AMD64

`# ##########################################################
#
# noso-go commit-e02c8066 by levi.noecker@gmail.com (c)2021
# https://github.com/leviable/noso-go
#
# ##########################################################
Connecting to Node1.mining.moe:8082 with password miningmoe
Using wallet address: gohosted
Number of CPU cores to use: 48
-> JOIN noso-go-commit-e02c8066
<- JOINOK N3N6LZVR342iyXgZA1rGvgEfQ1SDyCW "8!!!!!!! PoolData 9774 818FF842D5643B8D796C71AE84CE69A2 11 0 104 8013860 106 337450 3
-> PAYMENT
panic: runtime error: slice bounds out of range [:-1]

goroutine 1 [running]:
github.com/leviable/noso-go/internal/miner.parseAmount(0xc0002c0069, 0x7, 0x955d40, 0xc0001574d8)
        /Users/Levi/workspace/noso-go/internal/miner/payments.go:115 +0x1c5
github.com/leviable/noso-go/internal/miner.LogPaymentReq(0x741dce, 0x10, 0x7fffd3fc27b4, 0x8, 0x262e, 0xc0002c0069, 0x7)
        /Users/Levi/workspace/noso-go/internal/miner/payments.go:24 +0x6f
github.com/leviable/noso-go/internal/miner.Mine(0x955bc0)
        /Users/Levi/workspace/noso-go/internal/miner/mine.go:162 +0x1345
github.com/leviable/noso-go/cmd.glob..func4(0x94d3a0, 0xc000070aa0, 0x1, 0x5)
        /Users/Levi/workspace/noso-go/cmd/pool.go:107 +0x11e
github.com/spf13/cobra.(*Command).execute(0x94d3a0, 0xc000070a50, 0x5, 0x5, 0x94d3a0, 0xc000070a50)
        /Users/Levi/workspace/noso-go/vendor/github.com/spf13/cobra/command.go:856 +0x2c2
github.com/spf13/cobra.(*Command).ExecuteC(0x94d620, 0x41a501, 0x0, 0x0)
        /Users/Levi/workspace/noso-go/vendor/github.com/spf13/cobra/command.go:960 +0x375
github.com/spf13/cobra.(*Command).Execute(...)
        /Users/Levi/workspace/noso-go/vendor/github.com/spf13/cobra/command.go:897
github.com/leviable/noso-go/cmd.Execute()
        /Users/Levi/workspace/noso-go/cmd/root.go:63 +0x2d
main.main()
        /Users/Levi/workspace/noso-go/main.go:27 +0x25
`

--- a/.github/workflows/main.yaml
+++ b/.github/workflows/main.yaml
@@ -22,8 +22,16 @@ jobs:
       - name: Checkout master
         uses: actions/checkout@master
 
-      - name: Build binary
+      - name: Build ${{ matrix.binary }}
         run: make ${{ matrix.binary }}
+
+      - name: Compress ${{ matrix.binary }}
+        uses: crazy-max/ghaction-upx@v1
+        with:
+          version: latest
+          files: |
+            ./bin/${{ matrix.binary }}
+          args: -fq
 
       - uses: actions/upload-artifact@v2
         with:

--- a/.github/workflows/main.yaml
+++ b/.github/workflows/main.yaml
@@ -9,17 +9,26 @@ jobs:
   build:
     name: Build Binaries
     runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        binary: [noso-go-darwin-amd64,
+                 noso-go-linux-386,
+                 noso-go-linux-amd64,
+                 noso-go-linux-arm,
+                 noso-go-linux-arm64,
+                 noso-go-windows-386,
+                 noso-go-windows-amd64]
     steps:
       - name: Checkout master
         uses: actions/checkout@master
 
       - name: Build binary
-        run: make all
+        run: make ${{ matrix.binary }}
 
       - uses: actions/upload-artifact@v2
         with:
           path: |
-            bin/noso-go*
+            bin/${{ matrix.binary }}
             README.md
 
   test-linux:

--- a/Makefile
+++ b/Makefile
@@ -7,11 +7,14 @@ TAG := $(shell git describe --tags --exact-match $(REVISION) 2>/dev/null)
 
 ifneq ($(TAG),)
 VER := $(TAG)
+LDVER := -X 'github.com/leviable/noso-go/internal/miner.Version=$(VER)'
 else
-VER := commit-$(REVISION)
+LDVER :=
 endif
 
-LDFLAGS := -ldflags "-s -w -X 'github.com/leviable/noso-go/internal/miner.Version=$(VER)'"
+LDREV := -X 'github.com/leviable/noso-go/internal/miner.Commit=$(REVISION)'
+
+LDFLAGS := -ldflags "-s -w $(LDVER) $(LDREV)"
 
 .PHONY: all
 all: $(APP)-linux-amd64 $(APP)-linux-386 $(APP)-darwin-amd64 $(APP)-windows-386 $(APP)-windows-amd64 $(APP)-linux-arm $(APP)-linux-arm64

--- a/Makefile
+++ b/Makefile
@@ -11,7 +11,7 @@ else
 VER := commit-$(REVISION)
 endif
 
-LDFLAGS := -ldflags "-X 'github.com/leviable/noso-go/internal/miner.Version=$(VER)'"
+LDFLAGS := -ldflags "-s -w -X 'github.com/leviable/noso-go/internal/miner.Version=$(VER)'"
 
 .PHONY: all
 all: $(APP)-linux-amd64 $(APP)-linux-386 $(APP)-darwin-amd64 $(APP)-windows-386 $(APP)-windows-amd64 $(APP)-linux-arm $(APP)-linux-arm64
@@ -43,6 +43,7 @@ package-%:
 	@case $(OS) in \
 		linux) \
 			cp bin/$(APP)-$(OS)-$(ARCH) bin/$(APP);\
+			upx --brute bin/$(APP);\
 			cp examples/noso-go.sh bin/noso-go.sh;\
 			chmod +x bin/$(APP);\
 			chmod +x examples/noso-go.sh;\
@@ -51,6 +52,7 @@ package-%:
 			;; \
 		darwin) \
 			cp bin/$(APP)-$(OS)-$(ARCH) bin/$(APP);\
+			upx --brute bin/$(APP);\
 			cp examples/noso-go.sh bin/noso-go.sh;\
 			chmod +x bin/$(APP);\
 			chmod +x examples/noso-go.sh;\
@@ -58,6 +60,7 @@ package-%:
 			;; \
 		windows) \
 			cp bin/$(APP)-$(OS)-$(ARCH) bin/$(APP).exe;\
+			upx --brute bin/$(APP).exe;\
 			cp examples/$(APP).bat bin/$(APP).bat;\
 			chmod +x bin/$(APP).exe;\
 			if [ "$(ARCH)" = "386" ]; then \

--- a/Makefile
+++ b/Makefile
@@ -43,7 +43,6 @@ package-%:
 	@case $(OS) in \
 		linux) \
 			cp bin/$(APP)-$(OS)-$(ARCH) bin/$(APP);\
-			upx --brute bin/$(APP);\
 			cp examples/noso-go.sh bin/noso-go.sh;\
 			chmod +x bin/$(APP);\
 			chmod +x examples/noso-go.sh;\
@@ -52,7 +51,6 @@ package-%:
 			;; \
 		darwin) \
 			cp bin/$(APP)-$(OS)-$(ARCH) bin/$(APP);\
-			upx --brute bin/$(APP);\
 			cp examples/noso-go.sh bin/noso-go.sh;\
 			chmod +x bin/$(APP);\
 			chmod +x examples/noso-go.sh;\
@@ -60,7 +58,6 @@ package-%:
 			;; \
 		windows) \
 			cp bin/$(APP)-$(OS)-$(ARCH) bin/$(APP).exe;\
-			upx --brute bin/$(APP).exe;\
 			cp examples/$(APP).bat bin/$(APP).bat;\
 			chmod +x bin/$(APP).exe;\
 			if [ "$(ARCH)" = "386" ]; then \

--- a/cmd/mine.go
+++ b/cmd/mine.go
@@ -63,6 +63,7 @@ func init() {
 	mineCmd.Flags().StringVarP(&mineOpts.PoolPw, "password", "p", "", "Pool password")
 	mineCmd.Flags().StringVarP(&mineOpts.Wallet, "wallet", "w", "", "Noso wallet address to send payments to")
 	mineCmd.Flags().IntVarP(&mineOpts.Cpu, "cpu", "c", 4, "Number of CPU cores to use")
+	mineCmd.Flags().BoolVarP(&mineOpts.ShowPop, "show-pop", "", false, "Show PoP solutions in output")
 
 	mineCmd.MarkFlagRequired("address")
 	mineCmd.MarkFlagRequired("password")

--- a/cmd/mine.go
+++ b/cmd/mine.go
@@ -56,8 +56,6 @@ Example usage:
 func init() {
 	rootCmd.AddCommand(mineCmd)
 
-	// TODO: add --show-pop flag to show or not show PoP solutions
-	// TODO: add --show-step flag to show or not show STEP solutions
 	mineCmd.Flags().StringVarP(&mineOpts.IpAddr, "address", "a", "", "Pool IP address (e.g. 'noso.dukedog.io' or '75.45.193.238'")
 	mineCmd.Flags().IntVar(&mineOpts.IpPort, "port", 8082, "Pool port")
 	mineCmd.Flags().StringVarP(&mineOpts.PoolPw, "password", "p", "", "Pool password")

--- a/cmd/mine.go
+++ b/cmd/mine.go
@@ -56,6 +56,8 @@ Example usage:
 func init() {
 	rootCmd.AddCommand(mineCmd)
 
+	// TODO: add --show-pop flag to show or not show PoP solutions
+	// TODO: add --show-step flag to show or not show STEP solutions
 	mineCmd.Flags().StringVarP(&mineOpts.IpAddr, "address", "a", "", "Pool IP address (e.g. 'noso.dukedog.io' or '75.45.193.238'")
 	mineCmd.Flags().IntVar(&mineOpts.IpPort, "port", 8082, "Pool port")
 	mineCmd.Flags().StringVarP(&mineOpts.PoolPw, "password", "p", "", "Pool password")

--- a/cmd/pool.go
+++ b/cmd/pool.go
@@ -112,6 +112,7 @@ func init() {
 	poolCmd.Flags().BoolVarP(&info, "info", "i", false, "Print Pool information and exit")
 	poolCmd.Flags().StringVarP(&poolOpts.Wallet, "wallet", "w", "", "Noso wallet address to send payments to")
 	poolCmd.Flags().IntVarP(&poolOpts.Cpu, "cpu", "c", 4, "Number of CPU cores to use")
+	poolCmd.Flags().BoolVarP(&poolOpts.ShowPop, "show-pop", "", false, "Show PoP solutions in output")
 
 	poolCmd.Flags().SortFlags = false
 	poolCmd.Flags().PrintDefaults()
@@ -176,5 +177,10 @@ func loadPools() {
 		IpAddr: "Node1.mining.moe",
 		IpPort: 8082,
 		PoolPw: "miningmoe",
+	}
+	pools["devnoso"] = &miner.Opts{
+		IpAddr: "185.239.236.85",
+		IpPort: 8082,
+		PoolPw: "UnMaTcHeD",
 	}
 }

--- a/internal/miner/client.go
+++ b/internal/miner/client.go
@@ -12,7 +12,6 @@ const (
 	dialTimeout       = 5 * time.Second
 	connectionTimeout = 20 * time.Second
 	reconnectSleep    = 5 * time.Second
-	// reconnectSleep = 500 * time.Millisecond
 )
 
 func NewTcpClient(opts *Opts, comms *Comms) *TcpClient {

--- a/internal/miner/client.go
+++ b/internal/miner/client.go
@@ -32,6 +32,7 @@ type TcpClient struct {
 	RecvChan  chan string
 	conn      net.Conn
 	connected chan interface{}
+	manComms  *managerComms
 }
 
 type managerComms struct {
@@ -60,34 +61,35 @@ func NewManagerComms() *managerComms {
 func (t *TcpClient) manager() {
 	for {
 		t.conn = nil
-		manComms := NewManagerComms()
+		t.manComms = NewManagerComms()
 
-		go t.send(manComms)
-		go t.recv(manComms)
-		go t.ping(manComms)
+		go t.send()
+		go t.recv()
+		go t.ping()
+		go t.watchDog()
 
 		for running := true; running; {
 			select {
-			case <-manComms.disconnected:
-				close(manComms.stop)
+			case <-t.manComms.disconnected:
+				t.close(t.manComms.stop)
 				running = false
 			case <-t.comms.Joined:
-				close(manComms.joined)
+				t.close(t.manComms.joined)
 			}
 		}
 
 		// Wait for the goroutines to exit
 		// Should probably use a waitgroup here?
-		<-manComms.sendStopped
-		<-manComms.recvStopped
-		<-manComms.pingStopped
+		<-t.manComms.sendStopped
+		<-t.manComms.recvStopped
+		<-t.manComms.pingStopped
 
 		// Wait 5 seconds between connection attempts
-		time.Sleep(5 * time.Second)
+		time.Sleep(500 * time.Millisecond)
 	}
 }
 
-func (t *TcpClient) send(manComms *managerComms) {
+func (t *TcpClient) send() {
 	var (
 		err error
 	)
@@ -95,13 +97,13 @@ func (t *TcpClient) send(manComms *managerComms) {
 	if err != nil {
 		t.conn = nil
 		fmt.Printf("Error connecting to pool: %v\n", err)
-		close(manComms.disconnected)
-		close(manComms.sendStopped)
+		t.close(t.manComms.disconnected)
+		t.close(t.manComms.sendStopped)
 		return
 
 	} else {
 		t.conn.SetReadDeadline(time.Now().Add(20 * time.Second))
-		close(manComms.connected)
+		t.close(t.manComms.connected)
 	}
 
 	go t.join()
@@ -112,22 +114,22 @@ func (t *TcpClient) send(manComms *managerComms) {
 			fmt.Printf("-> %s\n", msg)
 			msg = fmt.Sprintf("%s %s\n", t.auth, msg)
 			fmt.Fprintf(t.conn, msg)
-		case <-manComms.stop:
-			close(manComms.sendStopped)
+		case <-t.manComms.stop:
+			t.close(t.manComms.sendStopped)
 			return
 		}
 	}
 }
 
-func (t *TcpClient) recv(manComms *managerComms) {
+func (t *TcpClient) recv() {
 	// Block until connection established
 	select {
-	case <-manComms.connected:
-	case <-manComms.stop:
-		close(manComms.recvStopped)
+	case <-t.manComms.connected:
+	case <-t.manComms.stop:
+		t.close(t.manComms.recvStopped)
 		return
 	}
-	<-manComms.connected
+	<-t.manComms.connected
 	scanner := bufio.NewScanner(t.conn)
 	for connected := true; connected; {
 		if ok := scanner.Scan(); !ok {
@@ -143,19 +145,19 @@ func (t *TcpClient) recv(manComms *managerComms) {
 		// Since we got something, reset the deadline
 		t.conn.SetReadDeadline(time.Now().Add(20 * time.Second))
 	}
-	close(manComms.disconnected)
-	close(manComms.recvStopped)
+	t.close(t.manComms.disconnected)
+	t.close(t.manComms.recvStopped)
 	return
 }
 
-func (t *TcpClient) ping(manComms *managerComms) {
+func (t *TcpClient) ping() {
 	var hashRate int
 
 	go func() {
 		for {
 			select {
 			case hashRate = <-t.comms.HashRate:
-			case <-manComms.stop:
+			case <-t.manComms.stop:
 				return
 			}
 		}
@@ -163,31 +165,57 @@ func (t *TcpClient) ping(manComms *managerComms) {
 
 	// Block until connected to pool
 	select {
-	case <-manComms.connected:
-	case <-manComms.stop:
-		close(manComms.pingStopped)
+	case <-t.manComms.connected:
+	case <-t.manComms.stop:
+		t.close(t.manComms.pingStopped)
 		return
 	}
 
 	// Block until pool has been joined
 	select {
-	case <-manComms.joined:
-	case <-manComms.stop:
-		close(manComms.pingStopped)
+	case <-t.manComms.joined:
+	case <-t.manComms.stop:
+		t.close(t.manComms.pingStopped)
 		return
 	}
 
 	for connected := true; connected; {
 		select {
-		case <-manComms.stop:
+		case <-t.manComms.stop:
 			connected = false
 		case <-time.After(5 * time.Second):
 			t.SendChan <- fmt.Sprintf("PING %d", hashRate/1000)
 		}
 	}
-	close(manComms.pingStopped)
+	t.close(t.manComms.pingStopped)
 }
 
 func (t *TcpClient) join() {
 	t.SendChan <- fmt.Sprintf("JOIN %s", t.minerVer)
+}
+
+func (t *TcpClient) close(c chan struct{}) {
+	select {
+	case <-c:
+		//fmt.Printf("********* Chan %+v already closed\n", c)
+	default:
+		//fmt.Printf("********* Closing Chan %+v \n", c)
+		close(c)
+	}
+}
+
+func (t *TcpClient) watchDog() {
+	// If we don't get a PONG back after X seconds, reconnect
+	for {
+		select {
+		case <-t.comms.Pong:
+			continue
+		case <-t.manComms.disconnected:
+			break
+		case <-time.After(20 * time.Second):
+			fmt.Printf("###################\nWatchdog Triggered\n###################\n")
+			t.close(t.manComms.disconnected)
+			break
+		}
+	}
 }

--- a/internal/miner/client.go
+++ b/internal/miner/client.go
@@ -4,7 +4,15 @@ import (
 	"bufio"
 	"fmt"
 	"net"
+	"sync"
 	"time"
+)
+
+const (
+	dialTimeout       = 5 * time.Second
+	connectionTimeout = 20 * time.Second
+	reconnectSleep    = 5 * time.Second
+	// reconnectSleep = 500 * time.Millisecond
 )
 
 func NewTcpClient(opts *Opts, comms *Comms) *TcpClient {
@@ -13,9 +21,10 @@ func NewTcpClient(opts *Opts, comms *Comms) *TcpClient {
 		comms:     comms,
 		addr:      fmt.Sprintf("%s:%d", opts.IpAddr, opts.IpPort),
 		auth:      fmt.Sprintf("%s %s", opts.PoolPw, opts.Wallet),
-		SendChan:  make(chan string, 0),
-		RecvChan:  make(chan string, 0),
+		SendChan:  make(chan string, 100),
+		RecvChan:  make(chan string, 100),
 		connected: make(chan interface{}, 0),
+		mutex:     &sync.Mutex{},
 	}
 
 	go client.manager()
@@ -32,190 +41,154 @@ type TcpClient struct {
 	RecvChan  chan string
 	conn      net.Conn
 	connected chan interface{}
-	manComms  *managerComms
+	mutex     *sync.Mutex
 }
 
 type managerComms struct {
 	connected    chan struct{}
 	disconnected chan struct{}
-	stop         chan struct{}
 	joined       chan struct{}
-	sendStopped  chan struct{}
-	recvStopped  chan struct{}
-	pingStopped  chan struct{}
 }
 
 func NewManagerComms() *managerComms {
 	return &managerComms{
 		connected:    make(chan struct{}, 0),
 		disconnected: make(chan struct{}, 0),
-		stop:         make(chan struct{}, 0),
 		joined:       make(chan struct{}, 0),
-		sendStopped:  make(chan struct{}, 0),
-		recvStopped:  make(chan struct{}, 0),
-		pingStopped:  make(chan struct{}, 0),
 	}
 }
 
 // Manages the TCP connection and send/recv/ping goroutines
 func (t *TcpClient) manager() {
 	for {
-		t.conn = nil
-		t.manComms = NewManagerComms()
+		manComms := NewManagerComms()
 
-		go t.send()
-		go t.recv()
-		go t.ping()
-		go t.watchDog()
+		conn, err := net.DialTimeout("tcp", t.addr, dialTimeout)
+		if err != nil {
+			fmt.Printf("Error connecting to pool: %v\n", err)
+		} else {
+			conn.SetReadDeadline(time.Now().Add(connectionTimeout))
 
-		for running := true; running; {
-			select {
-			case <-t.manComms.disconnected:
-				t.close(t.manComms.stop)
-				running = false
-			case <-t.comms.Joined:
-				t.close(t.manComms.joined)
+			go t.send(conn, manComms)
+			go t.recv(conn, manComms)
+			go t.ping(manComms)
+			go t.watchDog(manComms)
+
+		manager:
+			for {
+				select {
+				case <-manComms.disconnected:
+					break manager
+				case <-t.comms.Joined:
+					t.close(manComms.joined)
+				}
 			}
+
+			conn.Close()
 		}
 
-		// Wait for the goroutines to exit
-		// Should probably use a waitgroup here?
-		<-t.manComms.sendStopped
-		<-t.manComms.recvStopped
-		<-t.manComms.pingStopped
-
 		// Wait 5 seconds between connection attempts
-		time.Sleep(500 * time.Millisecond)
+		fmt.Printf("Disconnected from pool, will retry connection in %d seconds\n", reconnectSleep/time.Second)
+		time.Sleep(reconnectSleep)
 	}
 }
 
-func (t *TcpClient) send() {
-	var (
-		err error
-	)
-	t.conn, err = net.DialTimeout("tcp", t.addr, 5*time.Second)
-	if err != nil {
-		t.conn = nil
-		fmt.Printf("Error connecting to pool: %v\n", err)
-		t.close(t.manComms.disconnected)
-		t.close(t.manComms.sendStopped)
-		return
+func (t *TcpClient) send(conn net.Conn, manComms *managerComms) {
+	go func() { t.SendChan <- fmt.Sprintf("JOIN %s", t.minerVer) }()
 
-	} else {
-		t.conn.SetReadDeadline(time.Now().Add(20 * time.Second))
-		t.close(t.manComms.connected)
-	}
-
-	go t.join()
-
+send:
 	for {
 		select {
 		case msg := <-t.SendChan:
 			fmt.Printf("-> %s\n", msg)
 			msg = fmt.Sprintf("%s %s\n", t.auth, msg)
-			fmt.Fprintf(t.conn, msg)
-		case <-t.manComms.stop:
-			t.close(t.manComms.sendStopped)
-			return
+			fmt.Fprintf(conn, msg)
+		case <-manComms.disconnected:
+			break send
 		}
 	}
 }
 
-func (t *TcpClient) recv() {
-	// Block until connection established
-	select {
-	case <-t.manComms.connected:
-	case <-t.manComms.stop:
-		t.close(t.manComms.recvStopped)
-		return
-	}
-	<-t.manComms.connected
-	scanner := bufio.NewScanner(t.conn)
-	for connected := true; connected; {
-		if ok := scanner.Scan(); !ok {
-			fmt.Println("Error in connection: ", scanner.Err())
-			connected = false
+func (t *TcpClient) recv(conn net.Conn, manComms *managerComms) {
+	scanner := bufio.NewScanner(conn)
+recv:
+	for {
+		select {
+		case <-manComms.disconnected:
+			break recv
+		default:
+			if ok := scanner.Scan(); !ok {
+				fmt.Println("Error in connection: ", scanner.Err())
+				t.close(manComms.disconnected)
+				break
+			}
+			resp := scanner.Text()
+			if resp == "" {
+				continue
+			}
+			fmt.Print("<- " + resp + "\n")
+			t.RecvChan <- resp
+			// Since we got something, reset the deadline
+			conn.SetReadDeadline(time.Now().Add(connectionTimeout))
 		}
-		resp := scanner.Text()
-		if resp == "" {
-			continue
-		}
-		fmt.Print("<- " + resp + "\n")
-		t.RecvChan <- resp
-		// Since we got something, reset the deadline
-		t.conn.SetReadDeadline(time.Now().Add(20 * time.Second))
 	}
-	t.close(t.manComms.disconnected)
-	t.close(t.manComms.recvStopped)
-	return
 }
 
-func (t *TcpClient) ping() {
+func (t *TcpClient) ping(manComms *managerComms) {
 	var hashRate int
 
 	go func() {
 		for {
 			select {
-			case hashRate = <-t.comms.HashRate:
-			case <-t.manComms.stop:
+			case <-manComms.disconnected:
 				return
+			case hashRate = <-t.comms.HashRate:
 			}
 		}
 	}()
 
-	// Block until connected to pool
-	select {
-	case <-t.manComms.connected:
-	case <-t.manComms.stop:
-		t.close(t.manComms.pingStopped)
-		return
-	}
-
 	// Block until pool has been joined
 	select {
-	case <-t.manComms.joined:
-	case <-t.manComms.stop:
-		t.close(t.manComms.pingStopped)
+	case <-manComms.joined:
+	case <-manComms.disconnected:
 		return
 	}
 
-	for connected := true; connected; {
+ping:
+	for {
 		select {
-		case <-t.manComms.stop:
-			connected = false
+		case <-manComms.disconnected:
+			break ping
 		case <-time.After(5 * time.Second):
 			t.SendChan <- fmt.Sprintf("PING %d", hashRate/1000)
 		}
 	}
-	t.close(t.manComms.pingStopped)
 }
 
-func (t *TcpClient) join() {
-	t.SendChan <- fmt.Sprintf("JOIN %s", t.minerVer)
-}
-
-func (t *TcpClient) close(c chan struct{}) {
-	select {
-	case <-c:
-		//fmt.Printf("********* Chan %+v already closed\n", c)
-	default:
-		//fmt.Printf("********* Closing Chan %+v \n", c)
-		close(c)
-	}
-}
-
-func (t *TcpClient) watchDog() {
+func (t *TcpClient) watchDog(manComms *managerComms) {
 	// If we don't get a PONG back after X seconds, reconnect
+watchdog:
 	for {
 		select {
 		case <-t.comms.Pong:
 			continue
-		case <-t.manComms.disconnected:
-			break
-		case <-time.After(20 * time.Second):
+		case <-manComms.disconnected:
+			break watchdog
+		case <-time.After(connectionTimeout):
 			fmt.Printf("###################\nWatchdog Triggered\n###################\n")
-			t.close(t.manComms.disconnected)
-			break
+			t.close(manComms.disconnected)
+			break watchdog
 		}
+	}
+}
+
+func (t *TcpClient) close(c chan struct{}) {
+	t.mutex.Lock()
+	defer t.mutex.Unlock()
+
+	select {
+	case <-c:
+	default:
+		close(c)
 	}
 }

--- a/internal/miner/comms.go
+++ b/internal/miner/comms.go
@@ -11,6 +11,7 @@ func NewComms() *Comms {
 		Block:             make(chan int, 0),
 		Step:              make(chan int, 0),
 		Diff:              make(chan int, 0),
+		PoolDepth:         make(chan int, 0),
 		Balance:           make(chan string, 0),
 		BlocksTillPayment: make(chan int, 0),
 		StepSolved:        make(chan int, 0),
@@ -32,6 +33,7 @@ type Comms struct {
 	Block             chan int
 	Step              chan int
 	Diff              chan int
+	PoolDepth         chan int
 	Balance           chan string
 	BlocksTillPayment chan int
 	StepSolved        chan int

--- a/internal/miner/comms.go
+++ b/internal/miner/comms.go
@@ -20,6 +20,7 @@ func NewComms() *Comms {
 		Reports:           make(chan Report, 0),
 		Solutions:         make(chan Solution, 0),
 		Joined:            make(chan struct{}, 0),
+		Pong:              make(chan struct{}, 0),
 	}
 }
 
@@ -40,6 +41,7 @@ type Comms struct {
 	Reports           chan Report
 	Solutions         chan Solution
 	Joined            chan struct{}
+	Pong              chan struct{}
 }
 
 type Report struct {

--- a/internal/miner/comms.go
+++ b/internal/miner/comms.go
@@ -14,6 +14,7 @@ func NewComms() *Comms {
 		Balance:           make(chan string, 0),
 		BlocksTillPayment: make(chan int, 0),
 		StepSolved:        make(chan int, 0),
+		StepFailed:        make(chan int, 0),
 		HashRate:          make(chan int, 0),
 		Jobs:              make(chan Job, 0),
 		Reports:           make(chan Report, 0),
@@ -33,6 +34,7 @@ type Comms struct {
 	Balance           chan string
 	BlocksTillPayment chan int
 	StepSolved        chan int
+	StepFailed        chan int
 	HashRate          chan int
 	Jobs              chan Job
 	Reports           chan Report

--- a/internal/miner/jobs.go
+++ b/internal/miner/jobs.go
@@ -4,7 +4,19 @@ import (
 	"crypto/sha256"
 	"encoding/hex"
 	"fmt"
+	"math/rand"
 	"strings"
+	"time"
+)
+
+// ! is 33
+// % is 37
+// ( is 40
+// _ is 95
+// _ and ( are reserved chars, skip them
+// Wallet doesn't like %, skip it
+const (
+	hashableSeedChars = "!\"#$&')*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^`abcdefghijklmnopqrstuvwxyz{|"
 )
 
 func NewJobComms() *JobComms {
@@ -109,36 +121,22 @@ func JobFeeder(comms *Comms, jobComms *JobComms) {
 
 	<-ready
 
-	// ! is 33
-	// % is 37
-	// ( is 40
-	// _ is 95
-	// _ and ( are reserved chars, skip them
-	// Wallet doesn't like %, skip it
+	// Randomize seed chars so that if a miner restarts in the middle of a block,
+	// it isn't rehashing already hashed values
+	seedChars := []rune(hashableSeedChars)
+	rand.Seed(time.Now().UnixNano())
+	rand.Shuffle(len(seedChars), func(i, j int) { seedChars[i], seedChars[j] = seedChars[j], seedChars[i] })
 
 	for {
-		for x := 0; x < 92; x++ {
-			if x == 4 || x == 7 || x == 62 {
-				continue
-			}
-			for y := 0; y < 92; y++ {
-				if y == 4 || y == 7 || y == 62 {
-					continue
-				}
-				for z := 0; z < 92; z++ {
-					if z == 4 || z == 7 || z == 62 {
-						continue
-					}
-					seedBase := minerSeed[:len(minerSeed)-3]
-					seedX := string('!' + x)
-					seedY := string('!' + y)
-					seedZ := string('!' + z)
-					seed := seedBase + seedX + seedY + seedZ
+		for _, x := range seedChars {
+			for _, y := range seedChars {
+				for _, z := range seedChars {
 
-					// "_" and "(" are reserved characters in Noso
-					if strings.Contains(seed, "_") || strings.Contains(seed, "(") {
-						continue
-					}
+					seedBase := minerSeed[:len(minerSeed)-3]
+					seedX := string(x)
+					seedY := string(y)
+					seedZ := string(z)
+					seed := seedBase + seedX + seedY + seedZ
 
 				loop:
 					for num := 1; num < 999; num++ {

--- a/internal/miner/jobs.go
+++ b/internal/miner/jobs.go
@@ -16,6 +16,7 @@ func NewJobComms() *JobComms {
 		Diff:         make(chan int, 0),
 		TargetChars:  make(chan int, 0),
 		TargetString: make(chan string, 0),
+		PoolDepth:    make(chan int, 0),
 	}
 }
 
@@ -27,6 +28,7 @@ type JobComms struct {
 	Diff         chan int
 	TargetChars  chan int
 	TargetString chan string
+	PoolDepth    chan int
 }
 
 type Job struct {
@@ -40,6 +42,7 @@ type Job struct {
 	Diff          int
 	Block         int
 	Step          int
+	PoolDepth     int
 }
 
 func JobFeeder(comms *Comms, jobComms *JobComms) {
@@ -52,8 +55,8 @@ func JobFeeder(comms *Comms, jobComms *JobComms) {
 		targetChars  int
 		targetString string
 		job          Job
-		running      bool
 		postfix      string
+		poolDepth    int
 	)
 
 	verSha := sha256.Sum256([]byte(MinerName))
@@ -76,6 +79,7 @@ func JobFeeder(comms *Comms, jobComms *JobComms) {
 			case diff = <-jobComms.Diff:
 			case block = <-jobComms.Block:
 			case step = <-jobComms.Step:
+			case poolDepth = <-jobComms.PoolDepth:
 			}
 
 			if poolAddr == "" {
@@ -92,11 +96,13 @@ func JobFeeder(comms *Comms, jobComms *JobComms) {
 				continue
 			} else if targetString == "" {
 				continue
+			} else if poolDepth == 0 {
+				continue
 			} else if minerSeed == "" {
 				continue
 			} else {
-				ready <- struct{}{}
-				break
+				close(ready)
+				return
 			}
 		}
 	}()
@@ -134,15 +140,16 @@ func JobFeeder(comms *Comms, jobComms *JobComms) {
 						continue
 					}
 
+				loop:
 					for num := 1; num < 999; num++ {
 						postfix = ver + fmt.Sprintf("%03d", num)
-						running = true
 						fullSeed := seed + poolAddr + postfix
 						fullSeedBytes := []byte(fullSeed)
-						for running {
+						for {
 							job = Job{
 								TargetString:  strings.ToLower(targetString),
 								TargetChars:   targetChars,
+								PoolDepth:     poolDepth,
 								Diff:          diff,
 								Block:         block,
 								SeedMiner:     seed,
@@ -159,6 +166,8 @@ func JobFeeder(comms *Comms, jobComms *JobComms) {
 								job.TargetChars = targetChars
 							case targetString = <-jobComms.TargetString:
 								job.TargetString = targetString
+							case poolDepth = <-jobComms.PoolDepth:
+								job.PoolDepth = poolDepth
 							case diff = <-jobComms.Diff:
 								job.Diff = diff
 							case block = <-jobComms.Block:
@@ -166,7 +175,7 @@ func JobFeeder(comms *Comms, jobComms *JobComms) {
 							case step = <-jobComms.Step:
 								job.Step = step
 							case comms.Jobs <- job:
-								running = false
+								continue loop
 							}
 						}
 					}

--- a/internal/miner/mine.go
+++ b/internal/miner/mine.go
@@ -136,13 +136,16 @@ func Mine(opts *Opts) {
 			jobComms.TargetString <- targetString
 		case targetChars = <-comms.TargetChars:
 			jobComms.TargetChars <- targetChars
-		case targetBlock = <-comms.Block:
+		case newBlock := <-comms.Block:
+			if newBlock != targetBlock {
+				stepsAccepted += sumSteps(popSlice)
+				popSlice = make([]int, 0)
+				sharesEarned += sharesEarnedBlk
+				sharesEarnedBlk = 0
+			}
+			targetBlock = newBlock
 			jobComms.Block <- targetBlock
 			solComms.Block <- targetBlock
-			stepsAccepted += sumSteps(popSlice)
-			popSlice = make([]int, 0)
-			sharesEarned += sharesEarnedBlk
-			sharesEarnedBlk = 0
 		case currentStep = <-comms.Step:
 			jobComms.Step <- currentStep
 			solComms.Step <- currentStep
@@ -163,11 +166,13 @@ func Mine(opts *Opts) {
 			popSlice = append(popSlice, 0)
 			popCount++
 		case shares := <-comms.StepSolved:
-			popSlice[len(popSlice)-1]++
+			if len(popSlice) > 0 {
+				popSlice[len(popSlice)-1]++
+			}
 			sharesEarned += shares
 			sharesEarnedBlk += shares
 		case <-comms.StepFailed:
-			if popSlice[len(popSlice)-1] > 0 {
+			if len(popSlice) > 0 && popSlice[len(popSlice)-1] > 0 {
 				popSlice[len(popSlice)-1]--
 			}
 		case sol := <-comms.Solutions:

--- a/internal/miner/mine.go
+++ b/internal/miner/mine.go
@@ -28,10 +28,12 @@ import (
 )
 
 const (
-	HEADER = `# ##########################################################
+	HEADER = `
+# ##########################################################
 #
 # noso-go %s by levi.noecker@gmail.com (c)2021
 # https://github.com/leviable/noso-go
+# Commit: %s
 #
 # ##########################################################
 `
@@ -66,7 +68,7 @@ func Mine(opts *Opts) {
 		totalHashes int
 		hashRate    int
 	)
-	fmt.Printf(HEADER, Version)
+	fmt.Printf(HEADER, Version, Commit)
 
 	workerReports = make(map[string]Report)
 	popSlice = make([]int, 0)

--- a/internal/miner/mine.go
+++ b/internal/miner/mine.go
@@ -52,7 +52,11 @@ func Mine(opts *Opts) {
 		targetChars       int
 		currentStep       int
 		currentDiff       int
-		stepsSolved       int
+		popSlice          []int
+		popCount          int
+		stepsAccepted     int
+		sharesEarned      int
+		sharesEarnedBlk   int
 		blocksTillPayment int
 		balance           string
 		paymentRequested  time.Time
@@ -64,6 +68,7 @@ func Mine(opts *Opts) {
 	fmt.Printf(HEADER, Version)
 
 	workerReports = make(map[string]Report)
+	popSlice = make([]int, 0)
 
 	// Set a date in the past so we can request payment immediately if we
 	// have a vested balance
@@ -82,7 +87,7 @@ func Mine(opts *Opts) {
 
 	// Start the Solutions Manager goroutine
 	solComms := NewSolutionComms(client.SendChan)
-	go SolutionManager(solComms)
+	go SolutionManager(solComms, opts.ShowPop)
 
 	// Start the miner goroutines
 	ready := make(chan bool, 0)
@@ -101,6 +106,23 @@ func Mine(opts *Opts) {
 	// Create the payments.csv file if it doesn't already exist
 	CreateLogPaymentsFile()
 
+	// Print a reward status every 60 seconds
+	go func() {
+		for {
+			select {
+			case <-time.After(60 * time.Second):
+				fmt.Printf(
+					rewardMsg,
+					targetBlock,     // For Block %d
+					sharesEarnedBlk, // Shares Earned :
+					len(popSlice),   // PoP Sent :
+					sharesEarned,    // Shares Earned :
+					popCount,        // PoP Sent :
+				)
+			}
+		}
+	}()
+
 	// TODO: Sending individual info (block, chars, string, etc
 	//       will probably lead to a race condition. Send a
 	//       BlockUpdate struct instead with all info?
@@ -117,6 +139,10 @@ func Mine(opts *Opts) {
 		case targetBlock = <-comms.Block:
 			jobComms.Block <- targetBlock
 			solComms.Block <- targetBlock
+			stepsAccepted += sumSteps(popSlice)
+			popSlice = make([]int, 0)
+			sharesEarned += sharesEarnedBlk
+			sharesEarnedBlk = 0
 		case currentStep = <-comms.Step:
 			jobComms.Step <- currentStep
 			solComms.Step <- currentStep
@@ -133,9 +159,17 @@ func Mine(opts *Opts) {
 				LogPaymentReq(opts.IpAddr, opts.Wallet, targetBlock, balance)
 				paymentRequested = time.Now()
 			}
-		case <-comms.StepSolved:
-			stepsSolved += 1
-			fmt.Printf("Miner has solved %d steps\n", stepsSolved)
+		case <-solComms.StepSent:
+			popSlice = append(popSlice, 0)
+			popCount++
+		case shares := <-comms.StepSolved:
+			popSlice[len(popSlice)-1]++
+			sharesEarned += shares
+			sharesEarnedBlk += shares
+		case <-comms.StepFailed:
+			if popSlice[len(popSlice)-1] > 0 {
+				popSlice[len(popSlice)-1]--
+			}
 		case sol := <-comms.Solutions:
 			solComms.Solution <- sol
 		case report := <-comms.Reports:
@@ -154,3 +188,33 @@ func Mine(opts *Opts) {
 		}
 	}
 }
+
+func sumSteps(popSlice []int) (sum int) {
+	if len(popSlice) == 0 {
+		return 0
+	}
+	for _, v := range popSlice {
+		sum += v
+	}
+
+	return
+}
+
+const rewardMsg = `
+************************************
+
+Current Rewards Status
+
+For Block %d
+---------------
+Shares Earned  : %d
+PoP Sent       : %d
+
+Total
+---------------
+Shares Earned  : %d
+PoP Sent       : %d
+
+************************************
+
+`

--- a/internal/miner/mine.go
+++ b/internal/miner/mine.go
@@ -52,6 +52,7 @@ func Mine(opts *Opts) {
 		targetChars       int
 		currentStep       int
 		currentDiff       int
+		poolDepth         int
 		popSlice          []int
 		popCount          int
 		stepsAccepted     int
@@ -152,6 +153,8 @@ func Mine(opts *Opts) {
 		case currentDiff = <-comms.Diff:
 			jobComms.Diff <- currentDiff
 			solComms.Diff <- currentDiff
+		case poolDepth = <-comms.PoolDepth:
+			jobComms.PoolDepth <- poolDepth
 		case balance = <-comms.Balance:
 		case blocksTillPayment = <-comms.BlocksTillPayment:
 			// If we have a non-zero balance

--- a/internal/miner/miner.go
+++ b/internal/miner/miner.go
@@ -23,7 +23,6 @@ func Miner(workerNum string, comms *Comms, ready chan bool) {
 		targets     []string
 		targetLen   int
 		targetMin   int
-		poolDepth   int
 
 		// From hash_22
 		seedLen   int
@@ -35,8 +34,6 @@ func Miner(workerNum string, comms *Comms, ready chan bool) {
 		val       string
 		hashCount int
 	)
-
-	poolDepth = 2
 
 	encoded := make([]byte, 64)
 
@@ -50,14 +47,14 @@ func Miner(workerNum string, comms *Comms, ready chan bool) {
 	//   difficulty drops
 	for job := range comms.Jobs {
 		jobStart = time.Now()
-		targetMin = (job.Diff / 10) + 1 - poolDepth
+		targetMin = (job.Diff / 10) + 1 - job.PoolDepth
 		buff = bytes.NewBuffer(job.SeedFullBytes)
 		seedLen = buff.Len()
 		hashCount = 0
 
-		targets = make([]string, poolDepth+1)
+		targets = make([]string, job.PoolDepth+1)
 
-		for i := 0; i < poolDepth+1; i++ {
+		for i := 0; i < job.PoolDepth+1; i++ {
 			targets[i] = job.TargetString[:targetMin+i]
 		}
 

--- a/internal/miner/miner.go
+++ b/internal/miner/miner.go
@@ -4,6 +4,7 @@ import (
 	"bytes"
 	"crypto/sha256"
 	"encoding/hex"
+	"fmt"
 	"reflect"
 	"strings"
 	"time"
@@ -14,15 +15,16 @@ const (
 	hashChars = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
 )
 
-func Miner(worker_num string, comms *Comms, ready chan bool) {
+func Miner(workerNum string, comms *Comms, ready chan bool) {
 	var (
-		jobStart     time.Time
-		jobDuration  time.Duration
-		buff         *bytes.Buffer
-		hashStr      string
-		target_len   int
-		target_large string
-		target_small string
+		jobStart    time.Time
+		jobDuration time.Duration
+		buff        *bytes.Buffer
+		hashStr     string
+		targets     []string
+		targetLen   int
+		targetMin   int
+		poolDepth   int
 
 		// From hash_22
 		seedLen   int
@@ -35,21 +37,39 @@ func Miner(worker_num string, comms *Comms, ready chan bool) {
 		hashCount int
 	)
 
+	poolDepth = 3
+
 	encoded := make([]byte, 64)
 
 	// Wait until ready
 	<-ready
 
-	// Search for TargetChars - 1 solutions
-	// Report any TargetChars solutions immediately
-	// Store any TargetChars - 1 solutions until the steps drop
+	// - Search for TargetChars[:poolDepth] solutions
+	// - Report any TargetChars[:poolDepth] solutions immediately -> Proof of Participation
+	// - Report any TargetChars solutions immediately
+	// - Store any (TargetChars - 1) > poolDepth solutions until the step
+	//   difficulty drops
 	for job := range comms.Jobs {
 		jobStart = time.Now()
-		target_large = job.TargetString[:job.TargetChars]
-		target_small = job.TargetString[:job.TargetChars-1]
+		targetMin = job.TargetChars - poolDepth
 		buff = bytes.NewBuffer(job.SeedFullBytes)
 		seedLen = buff.Len()
 		hashCount = 0
+
+		targets = make([]string, poolDepth+1)
+
+		for i := 0; i < poolDepth+1; i++ {
+			targets[i] = job.TargetString[:targetMin+i]
+		}
+
+		// Sanity check: the last item in targets should be a full solution
+		// If it isn't, panic
+		if targets[len(targets)-1] != job.TargetString[:job.TargetChars] {
+			panic("Miner did not correctly build workload from Job")
+		}
+
+		// 5 was chosen so that it would take roughly 1 second to iterate
+		// through all the hashes on a modern-ish cpu thread
 		for _, w = range hashChars[:5] {
 			for _, x = range hashChars {
 				for _, y = range hashChars {
@@ -67,17 +87,25 @@ func Miner(worker_num string, comms *Comms, ready chan bool) {
 						hex.Encode(encoded, tmp[:])
 						val = BytesToString(encoded)
 
-						// We could almost certainly increase hashrate if we
-						// could search the sha sum bytes rather than converting
-						// to a string first and then doing a string search
-						// Also, need to benchmark doing a small substring search
-						if !strings.Contains(val, target_small) {
+						// TODO: We could almost certainly increase hashrate if we
+						//       could search the sha sum bytes rather than converting
+						//       to a string first and then doing a string search
+						// TODO: Benchmark doing a small substring search
+						if !strings.Contains(val, targets[0]) {
+							// targets[0] is that absolute minimum that a pool will accept
+							// if we dont match that minimum, we can drop this solution
+							// and continue with the hashing
 							continue
-						} else if strings.Contains(val, target_large) {
-							target_len = job.TargetChars
-						} else {
-							target_len = job.TargetChars - 1
 						}
+
+						targetLen = targetMin
+						for idx, t := range targets[1:] {
+							if !strings.Contains(val, t) {
+								break
+							}
+							targetLen = targetMin + idx
+						}
+						fmt.Println("****************\nFOUND ONE\n*****************")
 
 						hashStr = string(w) + string(x) + string(y) + string(z)
 						solution := make([]byte, len(val))
@@ -87,11 +115,11 @@ func Miner(worker_num string, comms *Comms, ready chan bool) {
 							Seed:       job.SeedMiner,
 							HashStr:    job.SeedPostfix + hashStr,
 							Block:      job.Block,
-							Chars:      target_len,
+							Chars:      targetLen,
 							Step:       job.Step,
 							SolvedHash: *(*string)(unsafe.Pointer(&solution)),
-							TargetLen:  target_len,
-							Target:     job.TargetString[:target_len],
+							TargetLen:  targetLen,
+							Target:     job.TargetString[:targetLen],
 							FullTarget: job.TargetString[:job.TargetChars],
 						}
 					}
@@ -99,7 +127,7 @@ func Miner(worker_num string, comms *Comms, ready chan bool) {
 			}
 		}
 		jobDuration = time.Since(jobStart)
-		comms.Reports <- Report{WorkerNum: worker_num, Hashes: hashCount, Duration: jobDuration}
+		comms.Reports <- Report{WorkerNum: workerNum, Hashes: hashCount, Duration: jobDuration}
 	}
 }
 

--- a/internal/miner/miner.go
+++ b/internal/miner/miner.go
@@ -61,12 +61,6 @@ func Miner(workerNum string, comms *Comms, ready chan bool) {
 			targets[i] = job.TargetString[:targetMin+i]
 		}
 
-		// Sanity check: the last item in targets should be a full solution
-		// If it isn't, panic
-		if targets[len(targets)-1] != job.TargetString[:job.TargetChars] {
-			panic("Miner did not correctly build workload from Job")
-		}
-
 		// 5 was chosen so that it would take roughly 1 second to iterate
 		// through all the hashes on one modern-ish cpu thread
 		for _, w = range hashChars[:5] {

--- a/internal/miner/miner.go
+++ b/internal/miner/miner.go
@@ -40,11 +40,6 @@ func Miner(workerNum string, comms *Comms, ready chan bool) {
 	// Wait until ready
 	<-ready
 
-	// - Search for TargetChars[:poolDepth] solutions
-	// - Report any TargetChars[:poolDepth] solutions immediately -> Proof of Participation
-	// - Report any TargetChars solutions immediately
-	// - Store any (TargetChars - 1) > poolDepth solutions until the step
-	//   difficulty drops
 	for job := range comms.Jobs {
 		jobStart = time.Now()
 		targetMin = (job.Diff / 10) + 1 - job.PoolDepth

--- a/internal/miner/miner.go
+++ b/internal/miner/miner.go
@@ -36,7 +36,7 @@ func Miner(workerNum string, comms *Comms, ready chan bool) {
 		hashCount int
 	)
 
-	poolDepth = 3
+	poolDepth = 2
 
 	encoded := make([]byte, 64)
 

--- a/internal/miner/miner.go
+++ b/internal/miner/miner.go
@@ -4,7 +4,6 @@ import (
 	"bytes"
 	"crypto/sha256"
 	"encoding/hex"
-	"fmt"
 	"reflect"
 	"strings"
 	"time"
@@ -51,7 +50,7 @@ func Miner(workerNum string, comms *Comms, ready chan bool) {
 	//   difficulty drops
 	for job := range comms.Jobs {
 		jobStart = time.Now()
-		targetMin = job.TargetChars - poolDepth
+		targetMin = (job.Diff / 10) + 1 - poolDepth
 		buff = bytes.NewBuffer(job.SeedFullBytes)
 		seedLen = buff.Len()
 		hashCount = 0
@@ -69,7 +68,7 @@ func Miner(workerNum string, comms *Comms, ready chan bool) {
 		}
 
 		// 5 was chosen so that it would take roughly 1 second to iterate
-		// through all the hashes on a modern-ish cpu thread
+		// through all the hashes on one modern-ish cpu thread
 		for _, w = range hashChars[:5] {
 			for _, x = range hashChars {
 				for _, y = range hashChars {
@@ -99,13 +98,12 @@ func Miner(workerNum string, comms *Comms, ready chan bool) {
 						}
 
 						targetLen = targetMin
-						for idx, t := range targets[1:] {
+						for _, t := range targets[1:] {
 							if !strings.Contains(val, t) {
 								break
 							}
-							targetLen = targetMin + idx
+							targetLen++
 						}
-						fmt.Println("****************\nFOUND ONE\n*****************")
 
 						hashStr = string(w) + string(x) + string(y) + string(z)
 						solution := make([]byte, len(val))
@@ -115,7 +113,7 @@ func Miner(workerNum string, comms *Comms, ready chan bool) {
 							Seed:       job.SeedMiner,
 							HashStr:    job.SeedPostfix + hashStr,
 							Block:      job.Block,
-							Chars:      targetLen,
+							Chars:      job.TargetChars,
 							Step:       job.Step,
 							SolvedHash: *(*string)(unsafe.Pointer(&solution)),
 							TargetLen:  targetLen,

--- a/internal/miner/opts.go
+++ b/internal/miner/opts.go
@@ -1,58 +1,10 @@
 package miner
 
-import (
-	"flag"
-	"fmt"
-	"os"
-	"runtime"
-)
-
 type Opts struct {
-	Cpu    int
-	IpAddr string
-	IpPort int
-	PoolPw string
-	Wallet string
-}
-
-func GetOpts() *Opts {
-	ver := flag.Bool("version", false, "Print version and exit")
-	// ipAddr := flag.String("addr", "", "IP address of Noso Pool")
-	// ipPort := flag.Int("port", 8082, "IP port of the Noso Pool. Defaults to 8082")
-	// poolPw := flag.String("password", "", "Password for the NosoPool")
-	wallet := flag.String("wallet", "", "Noso address from your wallet")
-	cpu := flag.Int("cpu", 0, "Number of CPUs to use when mining. Defaults to MAXCPU / 2")
-
-	flag.Parse()
-
-	if *ver {
-		fmt.Printf("Version: %s\n", Version)
-		os.Exit(0)
-		// } else if *ipAddr == "" {
-		// 	fmt.Println("-addr cannot be blank")
-		// 	os.Exit(1)
-		// } else if *poolPw == "" {
-		// 	fmt.Println("-password cannot be blank")
-		// 	os.Exit(1)
-	} else if *wallet == "" {
-		fmt.Println("-wallet cannot be blank")
-		os.Exit(1)
-	}
-
-	if *cpu == 0 {
-		c := runtime.NumCPU() / 2
-		cpu = &c
-	}
-
-	ipAddr := "noso.dukedog.io"
-	ipPort := 8082
-	poolPw := "duke"
-
-	return &Opts{
-		IpAddr: ipAddr,
-		IpPort: ipPort,
-		PoolPw: poolPw,
-		Wallet: *wallet,
-		Cpu:    *cpu,
-	}
+	Cpu     int
+	IpAddr  string
+	IpPort  int
+	PoolPw  string
+	Wallet  string
+	ShowPop bool
 }

--- a/internal/miner/parse.go
+++ b/internal/miner/parse.go
@@ -34,7 +34,7 @@ func Parse(comms *Comms, poolIp string, wallet string, block int, resp string) {
 	case PAYMENTOK:
 		LogPaymentResp(r, poolIp)
 	case PONG:
-		// NoOp
+		comms.Pong <- struct{}{}
 	case POOLSTEPS:
 		poolData(comms, r, 0)
 	case STEPOK:

--- a/internal/miner/parse.go
+++ b/internal/miner/parse.go
@@ -89,4 +89,11 @@ func poolData(comms *Comms, resp []string, offset int) {
 	} else {
 		comms.BlocksTillPayment <- blocksTillPayment
 	}
+
+	poolDepth, err := strconv.Atoi(resp[10+offset])
+	if err != nil {
+		fmt.Printf("Error converting target chars: %s\n", resp[10+offset])
+	} else {
+		comms.PoolDepth <- poolDepth
+	}
 }

--- a/internal/miner/parse.go
+++ b/internal/miner/parse.go
@@ -13,6 +13,7 @@ const (
 	PONG       = "PONG"
 	POOLSTEPS  = "POOLSTEPS"
 	STEPOK     = "STEPOK"
+	STEPFAIL   = "STEPFAIL"
 )
 
 func Parse(comms *Comms, poolIp string, wallet string, block int, resp string) {
@@ -37,8 +38,13 @@ func Parse(comms *Comms, poolIp string, wallet string, block int, resp string) {
 	case POOLSTEPS:
 		poolData(comms, r, 0)
 	case STEPOK:
-		fmt.Println("Step solution accepted by pool")
-		comms.StepSolved <- 1
+		shares, err := strconv.Atoi(r[1])
+		if err != nil {
+			fmt.Printf("Had trouble parsing the shares from STEPOK message: %v\n", err)
+		}
+		comms.StepSolved <- shares
+	case STEPFAIL:
+		comms.StepFailed <- 1
 	default:
 		fmt.Printf("Uknown response code: %s\n", r[0])
 	}

--- a/internal/miner/payments.go
+++ b/internal/miner/payments.go
@@ -111,10 +111,12 @@ func parseAmount(amount string) string {
 		// Already formatted to X.YYY, do nothing
 	} else if l == 8 {
 		amount = "0." + amount
-	} else {
+	} else if l > 8 {
 		whole = amount[:l-8]
 		frac = amount[l-8:]
 		amount = fmt.Sprintf("%s.%s", whole, frac)
+	} else {
+		amount = fmt.Sprintf("0.%08s", l)
 	}
 	return amount
 }

--- a/internal/miner/solution.go
+++ b/internal/miner/solution.go
@@ -73,7 +73,7 @@ func SolutionManager(solComms *SolutionComms) {
 				drop := 0
 				for idx, s := range sols {
 					if idx < 10 {
-						solComms.SendChan <- fmt.Sprintf("STEP %d %s %s", s.Block, s.Seed, s.HashStr)
+						solComms.SendChan <- fmt.Sprintf("STEP %d %s %s %d", s.Block, s.Seed, s.HashStr, s.Chars)
 					} else {
 						drop++
 					}
@@ -89,14 +89,14 @@ func SolutionManager(solComms *SolutionComms) {
 				// Drop stale solutions
 				fmt.Printf("Dropping Solution (old block): %+v\n", sol)
 				continue
-			} else if charsLrg-sol.Chars > 1 {
-				// Drop solutions where the step difficulty has already dropped
-				// and so these will never get sent. For instance, if diff is 95
-				// and we are on step 6, the miner will find solutions that are 8
-				// chars long, but those will never be valid
-				fmt.Printf("Dropping Solution (low diff): %+v\n", sol)
-				continue
-			} else if step < charsChg && sol.Chars < charsLrg {
+				// } else if charsLrg-sol.Chars > 1 {
+				// 	// Drop solutions where the step difficulty has already dropped
+				// 	// and so these will never get sent. For instance, if diff is 95
+				// 	// and we are on step 6, the miner will find solutions that are 8
+				// 	// chars long, but those will never be valid
+				// 	fmt.Printf("Dropping Solution (low diff): %+v\n", sol)
+				// 	continue
+			} else if step < charsChg && (sol.Chars+1) == charsLrg {
 				// Found a future solution, store it
 				// Should only store 10 max
 				// TODO: Max number of future solutions that can be accepted
@@ -112,7 +112,7 @@ func SolutionManager(solComms *SolutionComms) {
 				}
 			} else {
 				// Any solution that gets here is valid: send it
-				solComms.SendChan <- fmt.Sprintf("STEP %d %s %s", sol.Block, sol.Seed, sol.HashStr)
+				solComms.SendChan <- fmt.Sprintf("STEP %d %s %s %d", sol.Block, sol.Seed, sol.HashStr, sol.Chars)
 				if sol.Target == sol.FullTarget {
 					printFoundSolution(sol, false)
 				}

--- a/internal/miner/solution.go
+++ b/internal/miner/solution.go
@@ -34,15 +34,9 @@ type Solution struct {
 
 func SolutionManager(solComms *SolutionComms) {
 	var (
-		block      int
-		step       int
-		diff       int
-		charsCurr  int // Current chars, based on step and diff
-		charsChg   int // The step where char len change from lrg to sml
-		charsLrg   int // The largest possible char size for the diff
-		charsSml   int // The smallest possible char size for the diff
-		sol        Solution
-		storedSols map[int][]Solution
+		block int
+		diff  int
+		sol   Solution
 	)
 
 	for {
@@ -50,88 +44,42 @@ func SolutionManager(solComms *SolutionComms) {
 		case newBlock := <-solComms.Block:
 			// fmt.Println("Block is: ", block)
 			if newBlock != block {
-				storedSols = make(map[int][]Solution)
 				block = newBlock
 			}
-		case step = <-solComms.Step:
-			// fmt.Println("Step is: ", step)
+		case _ = <-solComms.Step:
 		case diff = <-solComms.Diff:
-			// fmt.Println("Diff is: ", diff)
-			// At diff 90, all steps have target chars 9
-			// At diff 95, steps 0 - 4 have target chars 10, 5 - 9 have 9
-			charsSml = diff / 10
-			charsLrg = charsSml + 1
-			charsChg = diff % 10
-			if step < charsChg {
-				charsCurr = charsLrg
-			} else {
-				charsCurr = charsSml
-			}
-
-			// Send any stored solutions that meet this criteria
-			if sols, ok := storedSols[charsCurr]; ok {
-				drop := 0
-				for idx, s := range sols {
-					if idx < 10 {
-						solComms.SendChan <- fmt.Sprintf("STEP %d %s %s %d", s.Block, s.Seed, s.HashStr, s.Chars)
-					} else {
-						drop++
-					}
-				}
-				if drop > 0 {
-					fmt.Println("Already sent 10 stored responses, dropping the remaining %d stored solutions")
-				}
-				storedSols[charsCurr] = make([]Solution, 0)
-			}
 		case sol = <-solComms.Solution:
 			fmt.Printf("Solution is: %+v\n", sol)
 			if sol.Block != block {
 				// Drop stale solutions
 				fmt.Printf("Dropping Solution (old block): %+v\n", sol)
 				continue
-				// } else if charsLrg-sol.Chars > 1 {
-				// 	// Drop solutions where the step difficulty has already dropped
-				// 	// and so these will never get sent. For instance, if diff is 95
-				// 	// and we are on step 6, the miner will find solutions that are 8
-				// 	// chars long, but those will never be valid
-				// 	fmt.Printf("Dropping Solution (low diff): %+v\n", sol)
-				// 	continue
-			} else if step < charsChg && (sol.Chars+1) == charsLrg {
-				// Found a future solution, store it
-				// Should only store 10 max
-				// TODO: Max number of future solutions that can be accepted
-				//       varies with the difficulty. Should only store max
-				//       possible and not just a blanket 10
-				if len(storedSols[sol.Chars]) < 10 {
-					storedSols[sol.Chars] = append(storedSols[sol.Chars], sol)
-					fmt.Println("Found a future solution, storing")
-					printFoundSolution(sol, true)
-					fmt.Printf("Currently have %d solutions stored\n", len(storedSols[sol.Chars]))
-				} else {
-					fmt.Printf("10 future solutions already stored, dropping: %+v\n", sol)
-				}
+			} else if sol.TargetLen <= sol.Chars-2 {
+				// PoP solution
+				printFoundSolution(sol, false)
+			} else if sol.TargetLen == sol.Chars-1 && diff%10 == 0 {
+				// PoP solution
+				printFoundSolution(sol, false)
+			} else if sol.TargetLen == sol.Chars-1 && diff%10 != 0 {
+				// Low step solution
+				printFoundSolution(sol, true)
 			} else {
-				// Any solution that gets here is valid: send it
-				solComms.SendChan <- fmt.Sprintf("STEP %d %s %s %d", sol.Block, sol.Seed, sol.HashStr, sol.Chars)
-				if sol.Target == sol.FullTarget {
-					printFoundSolution(sol, false)
-				}
+				// High step solution
+				printFoundSolution(sol, true)
 			}
+			solComms.SendChan <- fmt.Sprintf("STEP %d %s %s %d", sol.Block, sol.Seed, sol.HashStr, sol.TargetLen)
 		}
 	}
 }
 
-func printFoundSolution(sol Solution, future bool) {
-	currentOrFuture := "CURRENT"
-	storeOrSend := "Sending"
-	if future {
-		currentOrFuture = "FUTURE"
-		storeOrSend = "Storing"
+func printFoundSolution(sol Solution, isStep bool) {
+	stepOrPop := "PoP"
+	if isStep {
+		stepOrPop = "Step"
 	}
 	fmt.Printf(
 		found_one,
-		currentOrFuture,
-		storeOrSend,
+		stepOrPop,
 		sol.Block,
 		sol.Step,
 		sol.Seed,
@@ -145,7 +93,6 @@ func printFoundSolution(sol Solution, future bool) {
 
 const found_one = `************************************
 FOUND %s SOLUTION
-Store or Send : %s
 Block         : %d
 Step          : %d
 Seed          : %s

--- a/internal/miner/solution.go
+++ b/internal/miner/solution.go
@@ -9,6 +9,7 @@ func NewSolutionComms(sendChan chan string) *SolutionComms {
 		Diff:     make(chan int, 0),
 		Solution: make(chan Solution, 0),
 		SendChan: sendChan,
+		StepSent: make(chan struct{}, 0),
 	}
 }
 
@@ -18,6 +19,7 @@ type SolutionComms struct {
 	Diff     chan int
 	Solution chan Solution
 	SendChan chan string
+	StepSent chan struct{}
 }
 
 type Solution struct {
@@ -32,7 +34,7 @@ type Solution struct {
 	FullTarget string
 }
 
-func SolutionManager(solComms *SolutionComms) {
+func SolutionManager(solComms *SolutionComms, showPop bool) {
 	var (
 		block int
 		diff  int
@@ -49,17 +51,22 @@ func SolutionManager(solComms *SolutionComms) {
 		case _ = <-solComms.Step:
 		case diff = <-solComms.Diff:
 		case sol = <-solComms.Solution:
-			fmt.Printf("Solution is: %+v\n", sol)
+			// fmt.Printf("Solution is: %+v\n", sol)
 			if sol.Block != block {
 				// Drop stale solutions
 				fmt.Printf("Dropping Solution (old block): %+v\n", sol)
 				continue
 			} else if sol.TargetLen <= sol.Chars-2 {
 				// PoP solution
-				printFoundSolution(sol, false)
+				if showPop {
+					printFoundSolution(sol, false)
+				}
 			} else if sol.TargetLen == sol.Chars-1 && diff%10 == 0 {
 				// PoP solution
-				printFoundSolution(sol, false)
+				// When diff%10 == 0, there are no low steps
+				if showPop {
+					printFoundSolution(sol, false)
+				}
 			} else if sol.TargetLen == sol.Chars-1 && diff%10 != 0 {
 				// Low step solution
 				printFoundSolution(sol, true)
@@ -68,6 +75,7 @@ func SolutionManager(solComms *SolutionComms) {
 				printFoundSolution(sol, true)
 			}
 			solComms.SendChan <- fmt.Sprintf("STEP %d %s %s %d", sol.Block, sol.Seed, sol.HashStr, sol.TargetLen)
+			solComms.StepSent <- struct{}{}
 		}
 	}
 }

--- a/internal/miner/version.go
+++ b/internal/miner/version.go
@@ -3,6 +3,7 @@ package miner
 import "fmt"
 
 var (
+	Commit    = "0000000"
 	Version   = "v0.0.0"
-	MinerName = fmt.Sprintf("noso-go-%s", Version)
+	MinerName = fmt.Sprintf("ng%s", Version[1:])
 )

